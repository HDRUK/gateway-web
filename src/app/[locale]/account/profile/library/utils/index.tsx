import { IconButton } from "@mui/material";
import { createColumnHelper } from "@tanstack/react-table";
import { LibraryListItem, SelectedLibrary } from "@/interfaces/Library";
import Link from "@/components/Link";
import StyledCheckbox from "@/components/StyledCheckbox";
import TooltipIcon from "@/components/TooltipIcon";
import { CheckIcon, DeleteForeverIcon } from "@/consts/icons";
import { RouteName } from "@/consts/routeName";

const columnHelper = createColumnHelper<LibraryListItem>();

const getColumns = ({
    handleSelect,
    handleRemove,
    selected,
    translations,
}: {
    handleSelect: (data: SelectedLibrary) => void;
    handleRemove: (id: string | number) => void;
    selected: SelectedLibrary;
    translations: { [id: string]: string };
}) => [
    columnHelper.display({
        id: "actions",
        meta: { isPinned: true },
        cell: ({ row }) => {
            const {
                id,
                datasetId,
                name,
                dataCustodianId,
                dataCustodian,
                darEnabled,
                darTemplatePublished,
                cohortEnabled,
            } = row.original;
            return (
                <div style={{ textAlign: "center" }}>
                    <StyledCheckbox
                        checked={selected[id]?.selected}
                        onChange={(_e, value) =>
                            handleSelect({
                                [id]: {
                                    selected: value,
                                    datasetId,
                                    name,
                                    teamId: dataCustodianId,
                                    teamName: dataCustodian,
                                    darEnabled,
                                    darTemplatePublished,
                                    cohortEnabled,
                                },
                            })
                        }
                        size="large"
                        sx={{ p: 0 }}
                        iconSx={{ mr: 0 }}
                    />
                </div>
            );
        },
        header: () => null,
        size: 43,
    }),

    columnHelper.display({
        id: "name",
        cell: ({ row: { original } }) => {
            const { datasetId, name } = original;
            const linkHref = `/${RouteName.DATASET_ITEM}/${datasetId}`;

            return (
                <Link href={linkHref}>
                    <div style={{ textAlign: "center" }}>{name}</div>
                </Link>
            );
        },
        header: () => <span>{translations.name}</span>,
        size: 150,
    }),

    columnHelper.display({
        id: "darEnabled",
        cell: ({
            row: {
                original: { darEnabled },
            },
        }) => (
            <div style={{ textAlign: "center" }}>
                {darEnabled ? <CheckIcon color="primary" /> : "-"}
            </div>
        ),
        header: () => (
            <TooltipIcon
                buttonSx={{ p: 0 }}
                size="small"
                label={translations.darEnabled}
                content={translations.darEnabled}
            />
        ),
        size: 100,
    }),

    columnHelper.display({
        id: "cohortEnabled",
        cell: ({
            row: {
                original: { cohortEnabled },
            },
        }) => (
            <div style={{ textAlign: "center" }}>
                {cohortEnabled ? <CheckIcon color="primary" /> : "-"}
            </div>
        ),
        header: () => <span>{translations.cohortEnabled}</span>,
        size: 100,
    }),

    columnHelper.display({
        id: "dataCustodian",
        cell: ({
            row: {
                original: { dataCustodian },
            },
        }) => <div style={{ textAlign: "center" }}>{dataCustodian}</div>,
        header: () => <span>{translations.dataCustodian}</span>,
        size: 100,
    }),

    columnHelper.display({
        id: "entityType",
        cell: ({
            row: {
                original: { entityType },
            },
        }) => <div style={{ textAlign: "center" }}>{entityType}</div>,
        header: () => <span>{translations.entityType}</span>,
        size: 100,
    }),

    columnHelper.display({
        id: "delete",
        cell: ({ row: { original } }) => {
            return (
                <div style={{ textAlign: "center" }}>
                    <IconButton onClick={() => handleRemove(original.id)}>
                        <DeleteForeverIcon color="primary" />
                    </IconButton>
                </div>
            );
        },
        header: () => null,
        size: 43,
    }),
];

export { getColumns };
