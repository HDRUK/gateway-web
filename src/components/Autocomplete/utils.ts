import { ValueType } from "./Autocomplete";

const getChipLabel = (
    options: { value: string | number; label: string }[],
    value: ValueType
) => {
    if (typeof value === "string") {
        return value;
    }

    return options.find(option => option.value === value)?.label;
};

export { getChipLabel };
