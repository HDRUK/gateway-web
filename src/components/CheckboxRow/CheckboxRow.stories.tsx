import React from "react";
import { useForm } from "react-hook-form";
import { yupResolver } from "@hookform/resolvers/yup";
import { Stack } from "@mui/material";
import type { Meta, StoryObj } from "@storybook/react";
import * as yup from "yup";
import Button from "@/components/Button";
import CheckboxRow from "@/components/CheckboxRow";
import Form from "@/components/Form";

/** Mui documentation: https://mui.com/material-ui/react-checkbox */

const meta: Meta<typeof Form> = {
    component: Form,
    title: "Forms/CheckboxRow",
    tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof Form>;

export type FormData = {
    [key: string]: string;
};

const validationSchema = yup
    .object({
        second: yup.boolean().required().oneOf([true]),
    })
    .required();

const WrapperComponent = () => {
    const { handleSubmit, control } = useForm<FormData>({
        defaultValues: {
            first: "",
            second: "",
            third: "",
            fourth: "",
        },
        resolver: yupResolver(validationSchema),
    });

    const onSubmit = (data: unknown) => console.log(data);

    return (
        <Form onSubmit={handleSubmit(onSubmit)}>
            <Stack spacing={2} sx={{ marginBottom: 4, maxWidth: 400 }}>
                <CheckboxRow
                    label="Simple Checkbox"
                    control={control}
                    name="first"
                    title="Title"
                />
                <CheckboxRow
                    label="Required Checkbox"
                    required
                    control={control}
                    name="second"
                    title="Title"
                />
                <CheckboxRow
                    label="Disabled Checkbox"
                    disabled
                    control={control}
                    name="third"
                    title="Title"
                />
                <CheckboxRow
                    label="indeterminate Checkbox"
                    indeterminate
                    control={control}
                    name="fourth"
                    title="Title"
                />
                <Button type="submit">Submit</Button>
            </Stack>
        </Form>
    );
};

export const Default: Story = {
    render: () => <WrapperComponent />,
};
