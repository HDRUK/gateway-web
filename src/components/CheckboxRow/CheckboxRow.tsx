import Box from "@/components/Box";
import Checkbox from "@/components/Checkbox";
import { Typography } from "@mui/material";
import { CheckboxProps } from "../Checkbox/Checkbox";

export interface CheckboxRowProps extends CheckboxProps {
    title: string;
}

const CheckboxRow = ({ title, ...rest }: CheckboxRowProps) => {
    return (
        <Box
            sx={{
                display: "flex",
                alignItems: "center",
            }}>
            <Typography sx={{ width: "100px" }}>{title}</Typography>
            <Checkbox formControlSx={{ marginBottom: 0 }} {...rest} />
        </Box>
    );
};

export default CheckboxRow;
