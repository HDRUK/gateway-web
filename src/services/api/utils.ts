import { NotificationOptions } from "@/interfaces/Api";
import { AxiosResponse } from "axios";
import { Error } from "@/interfaces/Error";
import notificationService from "../notification";

interface ErrorNotificationProps {
    props: NotificationOptions;
    errorResponse: AxiosResponse<Error>;
    method: "delete" | "post" | "put" | "get";
}

const errorNotification = ({
    errorResponse,
    method,
    props,
}: ErrorNotificationProps) => {
    const { t, i18n, ...notificationProps } = props;
    const { data, status } = errorResponse || {};

    const fallbackTitle = i18n.exists(`api:common.error.status.${status}`)
        ? t(`api:common.error.status.${status}`)
        : "There has been an error";

    const title = data?.title || fallbackTitle;
    const message =
        data?.message ||
        t(`api:common.error.${method}.message`, {
            item: props.itemName || "Item",
        });

    notificationService.apiError(message, {
        title,
        message,
        errors: data?.errors,
        ...notificationProps,
    });
};

interface SuccessNotificationProps {
    props: NotificationOptions;
    method: "delete" | "post" | "put";
}

const successNotification = ({ props, method }: SuccessNotificationProps) => {
    const { t, i18n, ...notificationProps } = props;
    const customMessage = `api:${props.localeKey}.success.${method}.message`;
    const shouldOverideMessage = i18n.exists(customMessage);

    const message = shouldOverideMessage
        ? t(customMessage)
        : t(`api:common.success.${method}.message`, {
              item: props.itemName || "Item",
          });

    notificationService.success(message, { ...notificationProps });
};

export { errorNotification, successNotification };
