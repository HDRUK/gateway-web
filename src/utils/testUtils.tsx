import React, { ReactNode } from "react";
import {
    render,
    renderHook,
    RenderHookOptions,
    RenderOptions,
    RenderResult,
    RenderHookResult,
} from "@testing-library/react";
import { SWRConfig } from "swr";
import { CacheProvider, ThemeProvider } from "@emotion/react";
import ActionBarProvider from "@/providers/ActionBarProvider";
import DialogProvider from "@/providers/DialogProvider";
import theme from "@/config/theme";
import createEmotionCache from "@/config/createEmotionCache";

const clientSideEmotionCache = createEmotionCache();

const Wrapper = ({ children }: { children: ReactNode }) => {
    return (
        <SWRConfig
            value={{
                provider: () => new Map(),
            }}>
            <CacheProvider value={clientSideEmotionCache}>
                <ThemeProvider theme={theme}>
                    <ActionBarProvider>
                        <DialogProvider>{children}</DialogProvider>
                    </ActionBarProvider>
                </ThemeProvider>
            </CacheProvider>
        </SWRConfig>
    );
};

interface OptionProps
    extends RenderOptions<
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement,
        HTMLElement
    > {
    wrapperProps: Record<string, unknown>;
}

const customRender = (
    ui: React.ReactElement<
        unknown,
        string | React.JSXElementConstructor<unknown>
    >,
    options?: OptionProps
): RenderResult => {
    const { wrapperProps, ...rest } = options || {};

    return render(ui, {
        wrapper: props => <Wrapper {...props} {...wrapperProps} />,
        ...rest,
    });
};

interface OptionHookProps
    extends RenderHookOptions<
        unknown,
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement,
        HTMLElement
    > {
    wrapperProps: Record<string, unknown>;
}

const customRenderHook = (
    ui: (initialProps: unknown) => unknown,
    options?: OptionHookProps
): RenderHookResult<unknown, unknown> => {
    const { wrapperProps, ...rest } = options || {};
    return renderHook(ui, {
        wrapper: props => <Wrapper {...props} {...wrapperProps} />,
        ...rest,
    });
};

export * from "@testing-library/react";

export { customRenderHook as renderHook };
export { customRender as render };
