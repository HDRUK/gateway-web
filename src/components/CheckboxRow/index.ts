import CheckboxRow from "./CheckboxRow";

export default CheckboxRow;
