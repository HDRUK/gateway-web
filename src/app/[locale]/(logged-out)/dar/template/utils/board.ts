import { BoardSections, Board, Task } from "../components";
import { BOARD_SECTIONS } from "../constants";
import { getTasksByStatus } from "./tasks";

export const initializeBoard = (tasks: Task[]) => {
    const boardSections: BoardSections = {};

    Object.keys(BOARD_SECTIONS).forEach(boardSectionKey => {
        boardSections[boardSectionKey] = getTasksByStatus(
            tasks,
            boardSectionKey as Board
        );
    });

    return boardSections;
};

export const findBoardSectionContainer = (
    boardSections: BoardSections,
    id: string
) => {
    if (id in boardSections) {
        return id;
    }

    const container = Object.keys(boardSections).find(key =>
        boardSections[key].find(item => item.id === id)
    );
    return container;
};
