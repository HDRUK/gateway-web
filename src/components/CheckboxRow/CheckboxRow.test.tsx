import React from "react";
import CheckboxRow from "@/components/CheckboxRow";
import { useForm } from "react-hook-form";
import { render, screen } from "@/utils/testUtils";

describe("CheckboxRow", () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const Component = (props: any) => {
        const { control } = useForm();
        return (
            <CheckboxRow
                label="checkbox label"
                name="fieldName"
                title="Name"
                control={control}
                {...props}
            />
        );
    };
    it("should render component", async () => {
        const wrapper = render(<Component />);
        expect(wrapper.container).toMatchSnapshot();
    });

    it("should render title", async () => {
        render(<Component />);
        expect(screen.getByText("Name")).toBeInTheDocument();
    });
});
