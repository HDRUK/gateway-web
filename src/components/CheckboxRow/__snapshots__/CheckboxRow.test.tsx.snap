// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxRow should render component 1`] = `
<div>
  <div
    class="MuiBox-root mui-style-106jpu6"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 mui-style-12uetcm-MuiTypography-root"
    >
      Name
    </p>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth mui-style-2dhluk-MuiFormControl-root"
    >
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd mui-style-1mbnhdb-MuiFormControlLabel-root"
      >
        <span
          class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorSecondary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorSecondary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorSecondary MuiCheckbox-sizeMedium mui-style-1kn6ueq-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            class="PrivateSwitchBase-input mui-style-1m9pwf3"
            data-indeterminate="false"
            name="fieldName"
            type="checkbox"
          />
          <span
            class="mui-style-f6vn6k"
            font-size="medium"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label mui-style-1t8ylhr-MuiTypography-root"
        >
          checkbox label
        </span>
      </label>
    </div>
  </div>
</div>
`;
